src/storage/src/source/postgres/snapshot.rs

```diff
@@ -0,0 +1,376 @@
+) -> (
+    Collection<G, (usize, Result<Row, SourceReaderError>), Diff>,
+    Stream<G, RewindRequest>,
+    Stream<G, Rc<anyhow::Error>>,
+    Rc<dyn Any>,
+) {
+    let op_name = format!("TableReader({})", config.id);
+    let mut builder = AsyncOperatorBuilder::new(op_name, scope.clone());
+
+    let (feedback_handle, snapshot_feedback) = scope.feedback(Default::default());
+
+    let mut snapshot_input = builder.new_input(&snapshot_feedback, Pipeline);
+    let (mut raw_handle, raw_data) = builder.new_output_connection(vec![Antichain::new()]);
+    let (mut rewinds_handle, rewinds) = builder.new_output_connection(vec![Antichain::new()]);
+
+    // The export id must be sent to all workes, so we broadcast the feedback connection
+    let (mut snapshot_handle, snapshot) = builder.new_output_connection(vec![Antichain::new()]);
+    snapshot.broadcast().connect_loop(feedback_handle);
+
+    let is_snapshot_leader = config.responsible_for("snapshot_leader");
+
+    // A filtered table info containing only the tables that this worker should read
+    let mut reader_table_info = table_info.clone();
+    reader_table_info.retain(|oid, _| config.responsible_for(oid));
+    let (button, errors) = builder.build_fallible(move |caps| {
+        Box::pin(async move {
+            let id = config.id;
+            let worker_id = config.worker_id;
+
+            let [data_cap, rewind_cap, snapshot_cap]: &mut [_; 3] = caps.try_into().unwrap();
+            let data_cap = data_cap.as_mut().unwrap();
+            trace!(%id, "timely-{worker_id} initializing table reader {}", resume_upper.pretty());
+
+            if resume_upper.into_option() != Some(MzOffset::from(0)) {
+                return Ok(());
+            }
+
+            let connection_config = connection
+                .connection
+                .config(&*context.secrets_reader)
+                .await?
+                .replication_timeouts(config.params.pg_replication_timeouts.clone());
+
+            let client = connection_config.connect_replication().await?;
+            if is_snapshot_leader {
+                // The main slot must be created *before* we start snapshotting so that we can be
+                // certain that the temporarly slot created for the snapshot start at an LSN that
+                // is greater than or equal to that of the main slot.
+                super::ensure_replication_slot(&client, &connection).await?;
+
+                let snapshot_info = export_snapshot(&client).await?;
+                trace!(%id, "timely-{worker_id} exporting snapshot info {snapshot_info:?}");
+                let cap = snapshot_cap.as_ref().unwrap();
+                snapshot_handle.give(cap, snapshot_info).await;
+            }
+
+            let (snapshot, snapshot_lsn) = match snapshot_input.next_mut().await {
+                Some(AsyncEvent::Data(_, data)) => data.pop().expect("snapshot sent above"),
+                _ => unreachable!(),
```

Comment:
```suggestion
                _ => unreachable!("Data events always proceed Progress ones"),
```

Reply:
Good catch! I removed the `unreachable` because I don't want to depend on how the handle produces events